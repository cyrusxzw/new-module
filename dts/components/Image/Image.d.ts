import { FC } from 'react';
import type { ImageProps } from './Image.types';
declare const Image: FC<ImageProps>;
export { Image };
