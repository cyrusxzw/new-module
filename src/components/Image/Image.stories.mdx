import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { Image } from './Image';

<Meta title="Components/Image" component={Image} />

# Image

<Canvas>
  <Story name="Base component">
    <Image
      altText={knobs.text('altText', 'Lorem ipsum dolor sit amet')}
      isLazyLoaded={knobs.boolean('isLazyLoaded', true)}
      cta={{
        url: knobs.text('hyperlink.url', '/'),
      }}
      sizes={{
        large: knobs.text(
          'large',
          'https://www.aesop.com/medias/GelCleanser-640x640.png?context=bWFzdGVyfGltYWdlc3w1MzE5NTl8aW1hZ2UvcG5nfGltYWdlcy9oOTUvaGNkLzg4MDc3OTc4NTAxNDIucG5nfGM3NTNhYjdkMWNiOTRhNWZhNDA0NWExZWJiODMxN2VhNzEyY2ZkNGUyYzc4ZWVkZDRjYTIzMzZmZTRlOGYxNjM',
        ),
        medium: knobs.text(
          'medium',
          'https://www.aesop.com/medias/GelCleanser-640x640.png?context=bWFzdGVyfGltYWdlc3w1MzE5NTl8aW1hZ2UvcG5nfGltYWdlcy9oOTUvaGNkLzg4MDc3OTc4NTAxNDIucG5nfGM3NTNhYjdkMWNiOTRhNWZhNDA0NWExZWJiODMxN2VhNzEyY2ZkNGUyYzc4ZWVkZDRjYTIzMzZmZTRlOGYxNjM',
        ),
        small: knobs.text(
          'small',
          'https://www.aesop.com/medias/GelCleanser-640x640.png?context=bWFzdGVyfGltYWdlc3w1MzE5NTl8aW1hZ2UvcG5nfGltYWdlcy9oOTUvaGNkLzg4MDc3OTc4NTAxNDIucG5nfGM3NTNhYjdkMWNiOTRhNWZhNDA0NWExZWJiODMxN2VhNzEyY2ZkNGUyYzc4ZWVkZDRjYTIzMzZmZTRlOGYxNjM',
        ),
        xSmall: knobs.text('xSmall', ''),
        xLarge: knobs.text('xLarge', ''),
      }}
    />
  </Story>
</Canvas>

## API

<ArgsTable of={Image} />
