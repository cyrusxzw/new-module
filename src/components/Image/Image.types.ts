import { CSSProperties } from 'react';
import type { Themes } from '~/types';

type ImageSizeName = 'large' | 'medium' | 'small' | 'xLarge' | 'xSmall';
type ImageSizeCode = 'lg' | 'md' | 'sm' | 'xl' | 'xs';

type ImageProps = {
  altText?: string;
  className?: string;
  cta?: {
    openInANewWindow?: boolean;
    url?: string;
  };
  id?: string;
  isFullBleedImage?: boolean;
  isLazyLoaded?: boolean;
  sizes: {
    large?: string;
    medium?: string;
    small?: string;
    xLarge?: string;
    xSmall?: string;
  };
  style?: CSSProperties;
  type?: string;
  theme?: Themes;
};

export type { ImageProps, ImageSizeName, ImageSizeCode };
