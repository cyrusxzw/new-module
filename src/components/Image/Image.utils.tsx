import React from 'react';
import { BREAKPOINTS } from '~/constants';
import { uid } from 'react-uid';
import type { BreakpointNames } from '~/types';
import type { ImageProps, ImageSizeName, ImageSizeCode } from './Image.types';

const IMAGE_SIZES: Map<ImageSizeName, ImageSizeCode> = new Map([
  ['xSmall', 'xs'],
  ['small', 'sm'],
  ['medium', 'md'],
  ['large', 'lg'],
  ['xLarge', 'x3l'],
]);

function getReorderedSizes(sizes: ImageProps['sizes']) {
  /** Reorder props */
  const IMAGE_SOURCES: Map<ImageSizeName, string> = new Map([
    ['xLarge', sizes.xLarge],
    ['large', sizes.large],
    ['medium', sizes.medium],
    ['small', sizes.small],
    ['xSmall', sizes.xSmall],
  ]);

  return IMAGE_SOURCES;
}

function getImageSourcesBySize(
  sizes: ImageProps['sizes'],
): null | JSX.Element[] {
  if (!sizes) return null;
  /**
   * Ensure that there is a base / mobile first image if xSmall size prop is absent
   */
  let hasMobileFirstImage = false;

  if (!!Object.keys(sizes).indexOf('xSmall') && !!sizes['xSmall']) {
    hasMobileFirstImage = true;
  }

  const IMAGE_SOURCES = getReorderedSizes(sizes);
  const imageSources = [];

  IMAGE_SOURCES.forEach((source, name) => {
    const code: BreakpointNames | undefined =
      name === 'small' && !hasMobileFirstImage ? 'xs' : IMAGE_SIZES.get(name);

    if (!code || !source) return null;

    // eslint-disable-next-line
    console.log({ source, name, code });

    imageSources.push(
      <source
        key={uid(code)}
        media={`(min-width: ${BREAKPOINTS.get(code).minWidth}px)`}
        srcSet={source}
      />,
    );
  });

  // eslint-disable-next-line
  console.log({ imageSources, hasMobileFirstImage });

  return imageSources;
}

export { getImageSourcesBySize };
