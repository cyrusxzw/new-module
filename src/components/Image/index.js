import { Image } from './Image.tsx';

export { Image };
