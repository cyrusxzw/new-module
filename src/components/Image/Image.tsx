import React, { forwardRef, FC } from 'react';
import cx from 'classnames';
import get from 'lodash/get';
import { constraintsByViewport } from '~/utils/viewports';
import { Hyperlink } from '~/components/Hyperlink';
import type { ImageProps } from './Image.types';
import styles from './Image.module.css';

const Image: FC<ImageProps> = forwardRef(function ImageRef(
  {
    altText,
    className,
    cta,
    isLazyLoaded = true,
    large,
    medium,
    small,
    style,
    type,
    theme = 'dark',
  },
  ref: React.LegacyRef<HTMLImageElement>,
) {
  const fullBleedImage = type === 'componentFullBleedImage';
  const isRightAligned = type === 'componentSquareImage';

  const classSet = cx(
    { [styles.base]: !cta },
    styles.picture,
    { [styles.fullBleedImage]: fullBleedImage },
    { [styles.isRightAligned]: isRightAligned },
    styles[theme],
    className,
  );

  const picture = (
    <picture className={classSet} style={style}>
      {large && (
        <source
          media={`(min-width: ${get(
            constraintsByViewport,
            'large.minWidth',
          )}px)`}
          srcSet={large}
        />
      )}

      {medium && (
        <source
          media={`(min-width: ${get(
            constraintsByViewport,
            'medium.minWidth',
          )}px)`}
          srcSet={medium}
        />
      )}

      {small && (
        <source
          media={`(min-width: ${get(
            constraintsByViewport,
            'small.minWidth',
          )}px)`}
          srcSet={small}
        />
      )}

      <img alt={altText} loading={isLazyLoaded ? 'lazy' : 'eager'} ref={ref} />
    </picture>
  );

  return cta ? (
    <Hyperlink
      className={styles.base}
      hasTargetInNewWindow={cta.openInANewWindow}
      theme={theme}
      url={cta.url}
    >
      {picture}
    </Hyperlink>
  ) : (
    picture
  );
});

export { Image };
