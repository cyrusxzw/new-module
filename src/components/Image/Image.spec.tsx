import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';
import { Image } from './Image';

describe('<Image />', () => {
  it('should be defined', () => {
    expect(Image).toBeDefined();
  });

  it('should render base component correctly', () => {
    const { container } = render(
      <Image
        altText="Image of Milk Cleansing Campaign Aesop 2019"
        cta={{
          url: 'https://www.aesop.com/au/r/philosophy-to-products',
        }}
        large="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_large.png 2x"
        medium="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_medium.png 2x"
        small="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_small.png 2x"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders the picture tag without a Hyperlink if the `cta` prop was not passed', () => {
    const { container } = render(
      <Image
        altText="Image of Milk Cleansing Campaign Aesop 2019"
        large="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_large.png 2x"
        medium="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_medium.png 2x"
        small="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_small.png 2x"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders the picture without lazy loading', () => {
    const { container } = render(
      <Image
        altText="Image of Milk Cleansing Campaign Aesop 2019"
        isLazyLoaded={false}
        large="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_large.png 2x"
        medium="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_medium.png 2x"
        small="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_small.png 2x"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should be accessible', async () => {
    const { container } = render(
      <Image
        altText="Image of Milk Cleansing Campaign Aesop 2019"
        large="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_large.png 2x"
        medium="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_medium.png 2x"
        small="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_small.png 2x"
      />,
    );

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });
});
