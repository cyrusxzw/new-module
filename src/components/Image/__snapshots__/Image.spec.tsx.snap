// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> should render base component correctly 1`] = `
<div>
  <a
    class="base left dark base"
    href="https://www.aesop.com/au/r/philosophy-to-products"
    target="_self"
  >
    <picture
      class="picture dark"
    >
      <source
        media="(min-width: 1025px)"
        srcset="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_large.png 2x"
      />
      <source
        media="(min-width: 640px)"
        srcset="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_medium.png 2x"
      />
      <source
        media="(min-width: 0px)"
        srcset="/images/products/Hand_Reverence_Aromatique_Hand_Wash_500mL_small.png 2x"
      />
      <img
        alt="Image of Milk Cleansing Campaign Aesop 2019"
        loading="lazy"
      />
    </picture>
  </a>
</div>
`;
